tests/bsim/bluetooth/host/att/timeout/main.c

```diff
@@ -0,0 +1,288 @@
+	EXPECT_ZERO(bt_testlib_connect(&adva, &conn));
+
+	/* Establish EATT bearers. */
+	EXPECT_ZERO(bt_testlib_secure(conn, BT_SECURITY_L2));
+
+	while (bt_eatt_count(conn) == 0) {
+		k_msleep(100);
+	};
+
+	return conn;
+}
+
+static void central_read(struct bt_conn *conn, enum bt_att_chan_opt bearer, bool timeout)
+{
+	uint16_t actual_read_len;
+	uint16_t remote_read_send_len;
+	uint16_t handle = 0;
+	int err;
+
+	NET_BUF_SIMPLE_DEFINE(attr_value, sizeof(remote_read_send_len));
+
+	err = bt_testlib_att_read_by_type_sync(&attr_value, &actual_read_len, &handle, NULL, conn,
+					       bearer, UUID_2, BT_ATT_FIRST_ATTRIBUTE_HANDLE,
+					       BT_ATT_LAST_ATTRIBUTE_HANDLE);
+
+	if (timeout) {
+		__ASSERT(err == BT_ATT_ERR_UNLIKELY, "Unexpected error %d", err);
+	} else {
+		__ASSERT(!err, "Unexpected error %d", err);
+		__ASSERT(attr_value.len >= sizeof(remote_read_send_len),
+			 "Remote sent too little data.");
+		remote_read_send_len = net_buf_simple_pull_le16(&attr_value);
+		__ASSERT(remote_read_send_len == actual_read_len, "Length mismatch. %u %u",
+			 remote_read_send_len, actual_read_len);
+	}
+}
+
+static void wait_for_disconnect(bool timeout)
+{
+	int err;
+
+	err = k_sem_take(&disconnected_sem, K_SECONDS(BT_ATT_TIMEOUT_SEC + 2));
+
+	if (timeout) {
+		/* Here disconnect is triggered by the Central host due to ATT timeout. */
+		__ASSERT(!err, "Unexpected error %d", err);
+	} else {
+		/* Check that disconnect doesn't happen during time > ATT timeout. */
+		__ASSERT(err == -EAGAIN, "Unexpected error %d", err);
+	}
+}
+
+/**
+ * Test procedure:
+ *
+ * Central:
+ * 1. Connect to the peripheral.
+ * 2. Try to read a characteristic value.
+ * 3. Expect BT_ATT_ERR_UNLIKELY error.
+ * 4. Expect the peripheral to disconnect.
+ * 5. Reconnect to the peripheral.
+ * 6. Try to read a characteristic value.
+ * 7. Expect the peripheral to respond with the characteristic value.
+ * 8. Ensure that connection stays alive after a delay equal to ATT timeout.
+ * 9. Disconnect from the peripheral.
+ *
+ * Peripheral:
+ * 1. Start advertising.
+ * 2. Make the read callback sleep for more than ATT Timeout when the central tries to read.
+ * 3. Expect the disconnected callback to be called.
+ * 4. Start advertising again.
+ * 5. Make the read callback respond with the characteristic value when the central tries to read.
+ * 6. Expect the connection stay alive after a delay equal to ATT timeout.
+ * 7. Expect the central to disconnect.
+ */
+static void test_timeout(enum bt_att_chan_opt bearer)
+{
+	bool central = (get_device_nbr() == CENTRAL_DEVICE_NBR);
+	bool peripheral = (get_device_nbr() == PERIPHERAL_DEVICE_NBR);
+	struct bt_conn *conn;
+	int err;
+
+	/* Two iterations:
+	 * 1. ATT timeout.
+	 * 2. Successful read.
+	 */
+	for (int i = 0; i < 2; i++) {
+		bool timeout = i == 0;
+
+		if (peripheral) {
+			peripheral_setup(bearer, timeout);
+		}
+
+		if (central) {
+			conn = central_setup(bearer, timeout);
```

Comment:
This is overwriting a reference. Please initialize `conn` to NULL add `bt_testlib_conn_unref(&conn);` somewhere. Or perform a reference move to some other location.

Reply:
No point in setting `conn` to NULL in this function as it will be reassigned here. Removed `bt_conn_unref` from `disconnected` callback.